src/libcore/iter/mod.rs

```diff
@@ -1099,6 +1099,16 @@ impl<I: Iterator, P> Iterator for Filter<I, P> where P: FnMut(&I::Item) -> bool
         let (_, upper) = self.iter.size_hint();
         (0, upper) // can't know a lower bound, due to the predicate
     }
+
+    #[inline]
+    fn count(self) -> usize {
+        let (mut c, mut predicate, mut iter) = (0, self.predicate, self.iter);
+        for x in iter.by_ref() {
+            // branchless count
+            c += (&mut predicate)(&x) as usize;
+        }
+        c
```

Comment:
Good idea, will change shortly (perhaps also change the other methods end passant).